### DISCUSSION TITLE
Uploading files to S3

### DISCUSSION THREAD

**Discussion by edthetechie** [wheels-dev/wheels#1234] [Q&A] [answered]
I was wondering if anyone had any ideas or best practises for getting a file into S3 after it's uploaded onto the server via the CFWheels sendFile() function? I'm starting a new project and the client is asking for a CDN but I haven't worked with a CDN in Wheels before. I looked at the plugins but couldn't see one which relates to this so figured I'd ask.

Any help would be greatly appreciated :-)

**Comment by neokoenig** [comment, marked as answer]
I use this https://github.com/jcberquist/aws-cfml
I find it more useful that the built in CFML S3 stuff.

If you chuck that `miscellaneous/aws` you can do something like this:

```
// S3 CFC
local.aws = new app.miscellaneous.aws.aws(
    awsKey = YOURACCESSKEY,
    awsSecretKey = YOURSECRETKEY,
    defaultRegion = 'eu-west-2'
);

// S3 PUT
local.rv = local.aws.s3.putObject(
    bucket = BUCKETNAME,
    objectKey = "/uploads/1234/5678/1234.jpg"
    FileContent = uploadContent, // binary contents
    ContentType = 'image/jpeg' // A standard MIME type describing the format of the contents
);
```

**Reply by neokoenig** [reply]
PS SendFile() is for sending files to browser/user
// Send a PDF file to the user
sendFile(file="wheels_tutorial_20081028_J657D6HX.pdf");

If you're trying to then upload that file you'd need to read it into memory to send via the method above, via `FileReadBinary(file)`

**Reply by edthetechie** [reply]
Hey Tom, Thanks for such a quick reply. That looks like it will do the job very nicely, I've been using the AWS Java SDK to communicate with things like Amazon Polly, so that library could have other uses. Do you normally put your code in the controller or as an object callback in the model? I'm thinking probably the controller because the model doesn't know anything about a file being uploaded.

**Reply by edthetechie** [reply]
> PS SendFile() is for sending files to browser/user // Send a PDF file to the user sendFile(file="wheels_tutorial_20081028_J657D6HX.pdf");
> 
> If you're trying to then upload that file you'd need to read it into memory to send via the method above, via `FileReadBinary(file)`

Thanks for that, I actually knew sendFile() does that but it's late in the day and I'm running on empty :-) Thanks also for the explanation about FileReadBinary()

**Reply by neokoenig** [reply]
> Do you normally put your code in the controller or as an object callback in the model? I'm thinking probably the controller because the model doesn't know anything about a file being uploaded.

I call it from a controller, but actually use a tableless model to encapsulate the logic and config I need - I also run validation in it amongst other things. 

So in the end I just do 
```
local.s3upload = model("s3.fileUpload").new(...variousConfig);
if(local.s3upload.runValidation()) local.s3upload.doUpload();
```
Or there abouts. You get the idea.

**Reply by edthetechie** [reply]
That sounds like a really great way of doing it. I hadn't thought about the tableless model. Thank you so much for the help
